# Pirma Dalis

## 1.

In [1]:
# Importing packages
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

<font color='red'> Erikas Švažas: 

1. Norint išlaikyti formatavimą pagal PEP8 standartą labai padėtų papildomi extensions - pavyzdžiui nb_black
```
%load_ext nb_black
```
2. Prašymas kitam kartui - palikti kiekvienos celės outputą *visible*.
</font>

In [2]:
pd.set_option('display.max_rows', 500)

## 2. Parsisiųsti 2020 ir 2021 įmonių “Juridinių asmenų duomenis

In [3]:
# Reading data
# ba - balanso ataskaita
ba2020_data = pd.read_csv(
    'https://www.registrucentras.lt//aduomenys//?byla=JAR_FA_RODIKLIAI_BLNS_2020.csv')
ba2021_data = pd.read_csv(
    'https://www.registrucentras.lt//aduomenys//?byla=JAR_FA_RODIKLIAI_BLNS_2021.csv')
# pna - pelno/nuostoliu ataskaita
pna2020_data = pd.read_csv(
    'https://www.registrucentras.lt/aduomenys/?byla=JAR_FA_RODIKLIAI_PLNA_2020.csv')
pna2021_data = pd.read_csv(
    'https://www.registrucentras.lt/aduomenys/?byla=JAR_FA_RODIKLIAI_PLNA_2021.csv')

In [4]:
# copying datasets to reduce request count in case dataset collapses.
ba2020 = ba2020_data.copy()
ba2021 = ba2021_data.copy()
pna2020 = pna2020_data.copy()
pna2021 = pna2021_data.copy()

In [5]:
# setting names on dataframes
ba2020.name = 'ba2020'
ba2021.name = 'ba2021'
pna2020.name = 'pna2020'
pna2021.name = 'pna2021'

# 3. 3. Susipažinkite su duomenimis

## a. Ką reiškia skirtingi duomenų laukai? Iš kokių atributų sudaryti duomenys?

In [6]:
# ba dataframe attributes
ba_att = pd.read_excel(
    'https://www.registrucentras.lt/bylos/dokumentai/duomenys/aduomenys/JAR_FA_RODIKLIAI_BLNS_STR.xlsx', header=1)
ba_att

Žymėjimas    Tipas  \
0                      obj_kodas  decimal   
1                        obj_pav   string   
2                     form_kodas  decimal   
3                       form_pav   string   
4                  stat_statusas  decimal   
5                       stat_pav   string   
6                    template_id  decimal   
7                  template_name   string   
8                    standard_id  decimal   
9                  standard_name   string   
10               laikotarpis_nuo     date   
11               laikotarpis_iki     date   
12                      reg_date     date   
13            nuosavas_kapitalas  decimal   
14  mok_sumos_ir_isipareigojimai  decimal   
15             ilgalaikis_turtas  decimal   
16           trumpalaikis_turtas  decimal   

                                   Atributo aprašymas  
0   juridinio asmens kodas Juridinių asmenų registre   
1   juridinio asmens pavadinimas (individualių įmo...  
2                       juridinio asmens formos kodas  
3                 juridinio asmens formos pavadinimas  
4             juridinio asmens teisinio statuso kodas  
5       juridinio asmens teisinio statuso pavadinimas  
6                             ataskaitos rūšies kodas  
7                       ataskaitos rūšies pavadinimas  
8                         FA dokumentų rinkinio kodas  
9                         FA dokumentų rinkinio tipas  
10                      finansinių metų pradžios data  
11                      finansinių metų pabaigos data  
12                 FA dokumentų registravimo JAR data  
13  FA ataskaitos rodiklio (nuosavo kapitalo) reik...  
14  FA rodiklio (mokėtinų sumų ir įsipareigojimų) ...  
15  FA ataskaitos rodiklio (ilgalaikio turto) reik...  
16  FA ataskaitos rodiklio (trumpalaikio turto) re...

In [7]:
# pna dataframe attributes
pna_att = pd.read_excel(
    'https://www.registrucentras.lt/bylos/dokumentai/duomenys/aduomenys/JAR_FA_RODIKLIAI_PLNA_STR.xlsx', header=1)
pna_att

Žymėjimas    Tipas  \
0                    obj_kodas  decimal   
1                      obj_pav   string   
2                   form_kodas  decimal   
3                     form_pav   string   
4                stat_statusas  decimal   
5                     stat_pav   string   
6                  template_id  decimal   
7                template_name   string   
8                  standard_id  decimal   
9                standard_name   string   
10             laikotarpis_nuo     date   
11             laikotarpis_iki     date   
12                    reg_date     date   
13  pelnas_pries_apmokestinima  decimal   
14             grynasis_pelnas  decimal   
15           pardavimo_pajamos  decimal   

                                   Atributo aprašymas  
0   juridinio asmens kodas Juridinių asmenų registre   
1   juridinio asmens pavadinimas (individualių įmo...  
2                       juridinio asmens formos kodas  
3                 juridinio asmens formos pavadinimas  
4             juridinio asmens teisinio statuso kodas  
5       juridinio asmens teisinio statuso pavadinimas  
6                             ataskaitos rūšies kodas  
7                       ataskaitos rūšies pavadinimas  
8                         FA dokumentų rinkinio kodas  
9                         FA dokumentų rinkinio tipas  
10                      finansinių metų pradžios data  
11                      finansinių metų pabaigos data  
12                 FA dokumentų registravimo JAR data  
13  FA ataskaitos rodiklio (pelno prieš apmokestin...  
14  FA ataskaitos rodiklio (grynojo pelno) reikšmė...  
15  FA ataskaitos rodiklio (pardavimo pajamų) reik...

## b. Kiek reikšmių ir požymių turi kiekvienas iš dokumentų?

In [8]:
def get_features_and_values(df):
    name = df.name
    features = len(df.columns)
    values = len(df)
    return print(f'{name} dataset has:\n{features} features\n{values} values')

<font color='red'> Erikas Švažas: 

Šaunus fstring formatavimas! Be aware apie alternatyvas len() fcijai - pd.DataFrame.shape()
</font>

In [9]:
df_li = [ba2020, ba2021, pna2020, pna2021]

In [10]:
for df in df_li:
    get_features_and_values(df)
    print(' ')

ba2020 dataset has:
18 features
106318 values
 
ba2021 dataset has:
18 features
120850 values
 
pna2020 dataset has:
17 features
96616 values
 
pna2021 dataset has:
17 features
110227 values
 


## c. Kokio duomenų tipo yra kiekvienas iš požymių? Ar duomenų tipai tarp atitinkamų lentelių iš 2020 ir 2021 metų sutampa?


Duomenų tipus ir požymius galime matyti Pirmos dalies 3.a punkte.

<font color='red'> Erikas Švažas: 

*Pandas* infer duomenų tipą pagal pačioje *Pandas* esančius duomenų tipus atliekant *read_csv(). Sąlyga būtent ir prašo - koks duomenų tipas buvo priskirtas kiekvienam laukeliui. Šie tipai skiriasi nuo repositorijoje nurodytų duomenų tipų. [-0.3 taško]
</font>

## d. Paaiškinkite kas yra object duomenų tipas ir kuo jis skiriasi nuo string duomenų tipo.

Object duomenų tipas yra string arba string ir numeric verčių mišinys. pvz: data, kur turime skaičius ir brūkšnelius. Atpažinimo kodai kurie susideda iš raidžių ir skaičių. etc.

<font color='red'> Erikas Švažas: 

Dar pridėčiau, kad *everything is an object in Python* ir jei *Pandas* negali atpažinti duomenų tipo (ar tai integer ar string), by default bus *object*. Kiekvineas objektas turės skirtingas funkcijas ir savybes - pavyzdžiui norint naudoti string tipui būdingas fcijas .str.contains() reikės pradžioje atlikti konvertavimą į šį duomenų tipą.
</font> 

## e. Kuri lentelė turi daugiausiai nežinomų verčių? Kuris požymis išsiskiria nežinomų verčių skaičiumi?

In [16]:
# Iterating through frames to find out about nan values.
def Nan_info(df):
    name = df.name
    nanval = df.isna().sum().sum()
    maxval = df.isna().sum().sort_values(ascending=False)[:1][0]
    max_feat = df.isna().sum().sort_values(ascending=False)[:1].keys()[0]
    
    print(f'{name} dataset has {nanval} NaN values.\nFeature {max_feat} has the most Nan values,\nwith a total of {maxval}')

In [17]:
for df in df_li:
    Nan_info(df)
    print(' ')

ba2020 dataset has 17013 NaN values.
Feature mok_sumos_ir_isipareigojimai has the most Nan values,
with a total of 13644
 
ba2021 dataset has 24187 NaN values.
Feature mok_sumos_ir_isipareigojimai has the most Nan values,
with a total of 19670
 
pna2020 dataset has 47606 NaN values.
Feature pelnas_pries_apmokestinima has the most Nan values,
with a total of 30365
 
pna2021 dataset has 55858 NaN values.
Feature pelnas_pries_apmokestinima has the most Nan values,
with a total of 34755
 


Daugiausiai nežinomų verčių turi Pelno/Nuostolių ataskaitos.Požymis pelnas_pries_apmokestinima turi 55858 vertes, tiek 2020 tiek 2021 metais.


<font color='red'> Erikas Švažas: 

Kodas teisingas, tik interpretacija nėra tiksli - visas dataset'as turi ~56k trūkstamų verčių, bet požymis *pelnas_pries_apmokestinima* neturi 34755 verčių.

Patarimas - vietoje value_counts() galima naudoti 
````
dataframe.isna().sum().nlargest(1)
````
kad išsiaiškinti kuris feature turi daugiausiai missing values
</font> 

## f. Ar kažkuri lentelė turi pasikartojančių duomenų?

In [ ]:
# Checking for duplicates.
def has_dupes(df):
    name = df.name
    dupes = df.duplicated().sum()

    if dupes > 0:
        print(f'{name} has duplicates. Amount {dupes}')
    else:
        print(f'{name} does not have duplicate values.')

In [ ]:
for df in df_li:
    has_dupes(df)


<font color='red'> Erikas Švažas: 

Paprastesnė alternatyva:

```
ba2020[ba2020.duplicated()].shape

```

Nei viena lentelė neturi pasikartojančių duomenų.

## i. Kokia stulpelių aibė vienareikšmiškai identifikuoja kiekvienos lentelės eilutę?

In [ ]:
# Uniquity
ba2020[['obj_kodas','laikotarpis_nuo','laikotarpis_iki']].duplicated().sum()

Objekto kodas ir laikotarpis nuo iki parodo unikalumą šiame duomenų rinkinyje.

In [23]:
# Checking if dataframes are unique
def obj_nuo_iki_unique(df):
    name = df.name
    check = df[['obj_kodas','laikotarpis_nuo','laikotarpis_iki']].duplicated().sum()
    if check > 0:
        print(f"{name} dataset does not provide uniqueness through 'obj_kodas','laikotarpis_nuo','laikotarpis_iki'")
    else:
        print(f"{name} provides uniqueness through 'obj_kodas','laikotarpis_nuo','laikotarpis_iki'")

In [24]:
for df in df_li:
    obj_nuo_iki_unique(df)

ba2020 provides uniqueness through 'obj_kodas','laikotarpis_nuo','laikotarpis_iki'
ba2021 provides uniqueness through 'obj_kodas','laikotarpis_nuo','laikotarpis_iki'
pna2020 provides uniqueness through 'obj_kodas','laikotarpis_nuo','laikotarpis_iki'
pna2021 provides uniqueness through 'obj_kodas','laikotarpis_nuo','laikotarpis_iki'


<font color='red'> Erikas Švažas: 

Teisingai, bet iš esmės užtektų vieno laikotarpio stulpelio.
</font>

Kaip matome datasete objekto kodas su laikotarpiu rėžiu nusako unikalumą duomenų šaltinyje.

# Antra dalis

## 4. Keletas juridinių asmenų pavadinimų yra tiesiogiai susiję su “Thermo Fisher” bendrove. Raskite pilnus šių bendrovių pavadinimus.

<font color="red"> Erikas Švažas

Ateityje gali praversti .str.contains(case=?) argumentas

</font>

In [18]:
# names that match Thermo Fisher
ba2020[ba2020['obj_pav'].str.contains("Thermo Fisher")]['obj_pav'].unique()

array(['UAB "Thermo Fisher Scientific Vector Holdings"',
       'UAB "Thermo Fisher Scientific Baltics"',
       'UAB "Thermo Fisher Scientific Vector Senior Holdings I"',
       'UAB "Thermo Fisher Scientific Vector Senior Holdings II"'],
      dtype=object)

## 5. Apsimeskime, kad atributas “form_pav” neegzistuoja. Iš likusių laukelių gaukite visas UAB ir MB bendroves. Palyginkite rekonstrukcijos tikslumą su “form_pav” vertėmis - kiek procentų verčių pavyko atkurti skirtingose duomenyse?

In [ ]:
def get_form_name(name: str):
    name = name.lower()
    if 'uždaroji akcinė' in name or 'uab' in name:
        return 'Uždaroji akcinė bendrovė'
    if 'mb' in name:
        return 'Mažoji bendrija'
    if 'všį' in name or 'viešoji' in name:
        return 'Viešoji įstaiga'
    if 'labdaros' in name or 'fondas' in name:
        return 'Labdaros ir paramos fondas'
    return 'Asociacija'

In [ ]:
def get_reconstruct_percent(df):
    name = df.name
    length = len(df)  # Dataset length
    names = df['obj_pav'].apply(get_form_name) # Creating a series with modified name
    unmatched = len(df[~(df['form_pav'] == names)]) # Inverse search by name.
    
    percent = (unmatched/length)*100# Percent of matching error.
    
    print(f"{name} has a {round(percent,2)} % matching error with {unmatched} rows unmatched.")

In [ ]:
for df in df_li:
    get_reconstruct_percent(df)

<font color="red"> Erikas Švažas:

Puikiai, kad pastebėjote jog gali būti tiek UAB, tiek ilgasis pavadinimas (Uždaroji akcinė bendrovė). Taip pat čia buvo galima panaudoti str.split()/contains()... logiką ir nereikėtų išvardinti visų galimų if atvejų. 

Minusas - kai skaičiuojame rekonstrukcijos tikslumą turėtumę įtraukti tik exposed vertes (t.y. - mums neįdomi Viešoji įstaiga, bet įdomu tik UAB ir MB), tad reikėtų nusifiltruoti df_li vertes pagal obj_pav, kad išlaikytų tik UAB ir MB stulpelius (kuriuos turime) [-0.1 taško]
</font>

## 6. Apskaičiuoti vidutinias reikšmes “pelnas_pries_apmokestinima” ir “nuosavas_kapitalas” laukų kiekvienai juridinio asmens formai (Uždaroji akcinė bendrovė, Mažoji bendrija, etc.) už visą laikotarpį (2020 ir 2021 metai kartu)

In [ ]:
# Concatted dataset with 2020 and 2021 values.
pna = pd.concat([pna2020, pna2021], axis=0)
pna.reset_index(inplace=True)
pna['pelnas_pries_apmokestinima'].fillna(0, inplace=True)
pna.drop(columns='index', axis=0, inplace=True)
pna['reg_date'] = pd.to_datetime(pna['reg_date'])
ba = pd.concat([ba2020, ba2021], axis=0)
ba.reset_index(inplace=True)
ba.drop(columns='index', axis=0, inplace=True)
ba['nuosavas_kapitalas'].fillna(0, inplace=True)
ba['reg_date'] = pd.to_datetime(ba['reg_date'])

In [ ]:
# Unified the two datasets. Did not check for data leaks and breaks. (Probably a lot)
df = ba.merge(pna[['obj_kodas', 'laikotarpis_nuo', 'laikotarpis_iki', 'pelnas_pries_apmokestinima', 'grynasis_pelnas',
              'pardavimo_pajamos']], how='inner', on=['obj_kodas', 'laikotarpis_nuo', 'laikotarpis_iki'], indicator=False)
df.reset_index()
df['reg_date'] = pd.to_datetime(df['reg_date'])

In [ ]:
df.groupby('form_pav')[['pelnas_pries_apmokestinima']].mean()

In [ ]:
df.groupby('form_pav')[['nuosavas_kapitalas']].mean()

## 7. Atrinkti unikalias UAB ir MB įmones, kurių “nuosavas_kapitalas” buvo didesnis negu visų UAB ir MB įmonių nuosavo kapitalo mediana už visą laitotarpį (2020 ir 2021 metai kartu).

In [ ]:
#Getting median values for mb and uab firm types.
mb_median = df[df['form_pav'] == 'Mažoji bendrija'].groupby(
    'form_pav')['nuosavas_kapitalas'].median()[0]
uab_median = df[df['form_pav'] == 'Uždaroji akcinė bendrovė'].groupby(
    'form_pav')['nuosavas_kapitalas'].median()[0]

In [ ]:
temp7 = df[(df['form_pav'] == 'Mažoji bendrija') | (
    ba['form_pav'] == 'Uždaroji akcinė bendrovė')]

In [ ]:
temp7[(temp7['form_pav'] == 'Mažoji bendrija') & (temp7['nuosavas_kapitalas'] > mb_median) | (
    temp7['form_pav'] == 'Uždaroji akcinė bendrovė') & (temp7['nuosavas_kapitalas'] > uab_median)]

## 8. Įvertinkite kurios įmonės išsiskiria ypač didele “nuosavas_kapitalas” verte (patenka į 99th percentilę). Patikrinkite įmonių patenkančių į TOP-10 vertes rekvizitai.vz.lt svetainėje. Ranka įrašomi duomenys sukuria klaidos galimybę. Ar yra bendrovė, kurios “nuosavas_kapitalas” jums kelia įtarimų? 

In [ ]:
# Getting only companies that belong to 1% by 'nuosavas_kapitalas'
df[df['nuosavas_kapitalas'] > df['nuosavas_kapitalas'].quantile(0.99)].sort_values(by='nuosavas_kapitalas', ascending=False)

## 9. Vizualiai atvaizduoti juridinių asmenų formos pasiskirstymą per visą laikotarpį (2020 ir 2021 metais kartu). Vizualizacijos turi sutapti su pavaizduota dešinėje.

In [ ]:
# Pandas approach. Did not complete matplotlib and seaborn saddly.
df.groupby(['form_pav',df['reg_date'].dt.year])['obj_pav'].count().unstack().plot(kind='bar', logy=True, grid=True)
plt.title('Juridinių asmenų skaičius 2020 ir 2021 metais')
plt.ylabel('Vienetų skaičius')
plt.legend(['2020_count','2021_count'])

## 10. Kurį mėnesį buvo įregistruota daugiausiai juridinių asmenų 2020 ir 2021 metais? Atvaizduokite stulpelinę diagramą.

In [ ]:
df.groupby(['form_pav',df['reg_date'].dt.to_period('M')])['obj_pav'].count().unstack().sum().sort_values(ascending=False)[:2].plot(kind='bar')
plt.xlabel('Mėn.')
plt.ylabel('Juridinių asmenų sk.')

## 11. Vizualiai atvaizduoti juridinių asmenų statuso pasiskirstymą kiekvienais metais.

In [ ]:
df.groupby([df['reg_date'].dt.year, 'stat_pav'])['obj_pav'].count().unstack().plot(kind='bar', figsize=(16,8), logy=True, grid=True)
plt.legend()

## 12. Kiekvienai įmonei apskaičiuoti tendenciją. Jei “grynasis_pelnas” + “nuosavas_kapitalas” 2021 metais buvo didesnis už 2020 metais deklaruotą, tai tendencija “teigiama”, jei ne - “neigiama”. Atvaizduoti tendencijos duomenis stulpelinėje diagramoje.

In [ ]:
df['pajamos'] =  df['grynasis_pelnas'] + df['nuosavas_kapitalas']

In [ ]:
tend = pd.merge(
    df.loc[df['reg_date'].dt.year == 2020].groupby('obj_kodas')['pajamos'].sum(),
    df.loc[df['reg_date'].dt.year == 2021].groupby('obj_kodas')['pajamos'].sum(),
    how='inner',
    on='obj_kodas',
    suffixes=['2020', '2021']
)
tend.fillna(0, inplace=True)

In [ ]:
def tendency(x:int):
    if x > 0:
        return 'teigiama'
    return 'neigiama'

In [ ]:
tend['difference'] = tend['pajamos2020'] - tend['pajamos2021']
tend['tendency'] = tend['difference'].apply(tendency)
tend.reset_index(inplace=True)

In [ ]:
df = pd.merge(df,tend[['obj_kodas','tendency']],how='left',on='obj_kodas')

In [ ]:
df.groupby('tendency')['obj_kodas'].count().plot(kind='bar')
plt.xlabel('Tendencija')

## 13. Surasti UAB ir MB įmones, kurios 2020 vykdė veiklą, o 2021 - bankrutavo. Bankrutavusių įmonių statusai: “Bankrutavęs”, “Bankrutuojantis”, “Išregistruotas”, “Inicijuojamas likvidavimas”, “Likviduojamas”, “Likviduojamas dėl bankroto”. Toliau veiklą vykdanti įmonė pažymėta "Teisinis stat neįregistruotas" statusu. Kokia bankrutavusių įmonių “nuosavas_kapitalas”suma 2021 metais? Kokia bankrutavusių įmonių “grynasis_pelnas” suma 2021 metais? Palyginti šias sumas su veikiančių įmonių atitinkamomis sumomis.

In [ ]:
uabmb = df[df['form_pav'].isin(['Uždaroji akcinė bendrovė','Mažoji bendrija'])]

In [ ]:
# bankrupt filter
bankrupt = ['Išregistruotas', 'Likviduojamas',
            'Likviduojamas dėl bankroto', 'Bankrutuojantis',
            'Inicijuojamas likvidavimas', 'Bankrutavęs'] 
# companies working in 2020
working = uabmb.loc[(uabmb['reg_date'].dt.year == 2020) & (
    uabmb['stat_pav'].isin(['Teisinis stat neįregistruotas']))]['obj_kodas']

In [ ]:
# Bankrupt company codes in 2021
bankrupted = uabmb.loc[(uabmb['reg_date'].dt.year == 2021) & (
    uabmb['stat_pav'].isin(bankrupt))][['obj_kodas', 'stat_pav']]

In [ ]:
# Tried merging with inner to find companies that worked in 2020 but are bankrupt in 2021.
pd.merge(working,bankrupted,how='left',on='obj_kodas')

## 14. Sukurti naują stulpelį “turtas”, kuris skaičiuojamas sudedant “trumpalaikis_turtas” ir “ilgalaikis_turtas”

In [ ]:
df['trumpalaikis_turtas'].fillna(0, inplace=True)
df['ilgalaikis_turtas'].fillna(0, inplace=True)

In [ ]:
df['turtas'] = df['trumpalaikis_turtas'] + df['ilgalaikis_turtas']

In [ ]:
bins = [0, 3000, 20_000, 35_000, 70_000, df['turtas'].max()]

In [ ]:
df['grupe'] =  pd.cut(df['turtas'], bins=bins)

In [ ]:
fig, axes = plt.subplots(1,2, figsize=(18,8), sharey=True)
labels = ['iki 3k', '3k-20k', '20k-35k', '35k-70k', '70k+'] # Labels for pie chart
df[df['reg_date'].dt.year == 2020].groupby([df['reg_date'].dt.year, df['grupe']])[
    'obj_kodas'].count().plot(kind='pie', autopct='%1.1f%%', ax=axes[0], labels=labels, startangle=145, ylabel='', xlabel='')
df[df['reg_date'].dt.year == 2021].groupby([df['reg_date'].dt.year, df['grupe']])[
    'obj_kodas'].count().plot(kind='pie', autopct='%1.1f%%', ax=axes[1], labels=labels, startangle=145)
plt.suptitle('Įregistruotų įmonių pasiskirstymas pagal sukauptą turtą') # Title for figure not subplots
plt.legend(labels) # Adding legend

## 15. Rasti TOP-100 pelningiausių įmonių pagal “grynasis pelnas” ir TOP-100 įmonių kurių “nuosavas kapitalas” didžiausias. Ar yra sutampančių įmonių šiuose sąrašuose? Skaičiuojant TOP įmonės įtraukti viso laikotarpio įmones (naudoti 2020 ir 2021 metų duomenis kartu).

In [ ]:
#TOP 100 by income.
top100gp = df.groupby('obj_pav')[['grynasis_pelnas']].sum().sort_values(by='grynasis_pelnas', ascending=False).reset_index()[:100]

In [ ]:
#TOP 100 by capital
top100nk = df.groupby('obj_pav')[['nuosavas_kapitalas']].sum().sort_values(by='nuosavas_kapitalas', ascending=False).reset_index()[:100]

In [ ]:
# Company count that shows up in both
len(top100gp[top100gp['obj_pav'].isin(top100nk['obj_pav'].tolist())])

## 16. Ar TOP įmonės pagal “grynasis pelnas” iš 15 užduoties skiriasi pagal metus, t.y. 2020 ir 2021 metais? Kurios įmonės iškrenta iš TOP sąrašo 2021 metais?

In [ ]:
#Top 100 income 2020 and 2021
top100gp2020 = df[df['reg_date'].dt.year == 2020].groupby('obj_pav')[['grynasis_pelnas']].sum().sort_values(by='grynasis_pelnas', ascending=False).reset_index()[:100]
top100gp2021 = df[df['reg_date'].dt.year == 2021].groupby('obj_pav')[['grynasis_pelnas']].sum().sort_values(by='grynasis_pelnas', ascending=False).reset_index()[:100]

In [ ]:
# Companies that drop out in 2021
top20butnot21 = top100gp2020[~top100gp2020.isin(top100gp2021['obj_pav'].tolist())].dropna()
top20butnot21[['obj_pav']]

# Trečia dalis

<font color='red'> Justas Kalpokas: </font>
1. <font color='red'> paversti i data butu efektyviau isparsinant visa stulpeli, o ne iteruojant po viena. 
pvz.
```
 pd.to_datetime(sod20['date'].map(lambda x: x[:4]+ "-"+ x[-2:]))
 ```
<font color='red'> 
1. <font color='red'> sukurete 3 augimo kriterijus. Atlyginimas, pardavimai ir darbuotoju skaicius. Manau reikejo prideti svorius ir is visu kriteriju ieskoti kombinacijos, kuri leistu surikiuoti imones.    <font color='red'> 





## 1. Jums prieinamą finansinę informaciją papildykite

In [ ]:
# Column mapper by task
cols = pd.read_csv('data/3_uzduotis_data_sample.csv').columns

In [ ]:
# Bandymas su apply pasidaryti metų columns, tačiau labai ilgai krovėsi.
#def get_2020_nuosavas_kapitalas(obj:int):
#    getsum = float(df[(df['obj_kodas'] == obj) & (df['reg_date'].dt.year == 2020)].groupby('obj_kodas')['nuosavas_kapitalas'].sum())
#    return getsum

In [ ]:
# Bruteforce is better than nothing.
features = [
    df[df['reg_date'].dt.year == 2020].groupby('obj_kodas')[['nuosavas_kapitalas']].sum(
    ).rename(columns={'nuosavas_kapitalas': 'nuosavas_kapitalas_20'}).reset_index()
    ,df[df['reg_date'].dt.year == 2021].groupby('obj_kodas')[['nuosavas_kapitalas']].sum(
    ).rename(columns={'nuosavas_kapitalas': 'nuosavas_kapitalas_21'}).reset_index()
    ,df[df['reg_date'].dt.year == 2020].groupby('obj_kodas')[['mok_sumos_ir_isipareigojimai']].sum(
    ).rename(columns={'mok_sumos_ir_isipareigojimai': 'mok_sumos_ir_isipareigojimai_20'}).reset_index()
    ,df[df['reg_date'].dt.year == 2021].groupby('obj_kodas')[['mok_sumos_ir_isipareigojimai']].sum(
    ).rename(columns={'mok_sumos_ir_isipareigojimai': 'mok_sumos_ir_isipareigojimai_21'}).reset_index()
    ,df[df['reg_date'].dt.year == 2020].groupby('obj_kodas')[['ilgalaikis_turtas']].sum(
    ).rename(columns={'ilgalaikis_turtas': 'ilgalaikis_turtas_20'}).reset_index()
    ,df[df['reg_date'].dt.year == 2021].groupby('obj_kodas')[['ilgalaikis_turtas']].sum(
    ).rename(columns={'ilgalaikis_turtas': 'ilgalaikis_turtas_21'}).reset_index()
    ,df[df['reg_date'].dt.year == 2020].groupby('obj_kodas')[['trumpalaikis_turtas']].sum(
    ).rename(columns={'trumpalaikis_turtas': 'trumpalaikis_turtas_20'}).reset_index()
    ,df[df['reg_date'].dt.year == 2021].groupby('obj_kodas')[['trumpalaikis_turtas']].sum(
    ).rename(columns={'trumpalaikis_turtas': 'trumpalaikis_turtas_21'}).reset_index()
    ,df[df['reg_date'].dt.year == 2020].groupby('obj_kodas')[['pardavimo_pajamos']].sum(
    ).rename(columns={'pardavimo_pajamos': 'pardavimo_pajamos_20'}).reset_index()
    ,df[df['reg_date'].dt.year == 2021].groupby('obj_kodas')[['pardavimo_pajamos']].sum(
    ).rename(columns={'pardavimo_pajamos': 'pardavimo_pajamos_21'}).reset_index()
    ,df[df['reg_date'].dt.year == 2020].groupby('obj_kodas')[['grynasis_pelnas']].sum(
    ).rename(columns={'grynasis_pelnas': 'grynasis_pelnas_20'}).reset_index()
    ,df[df['reg_date'].dt.year == 2021].groupby('obj_kodas')[['grynasis_pelnas']].sum(
    ).rename(columns={'grynasis_pelnas': 'grynasis_pelnas_21'}).reset_index()
]

In [ ]:
# Merging original df with subdataframes
for feat in features:
    df = pd.merge(df,feat,on='obj_kodas',how='left')

In [ ]:
# Neleidžia nuskaityti iš interneto, HTTP request 403
#pd.read_csv('https://atvira.sodra.lt/imones/downloads/2018/monthly-2018.csv.zip', compression='zip')

Atsiunčiami 2020 metų duomenys.

In [ ]:
# Copying dataframe in case of breaking.
sod = df.copy()

In [ ]:
# Remapping verbose feature names into more concise.
sod_map = {
    'Draudėjo kodas (code)':'code',
    'Juridinių asmenų registro kodas (jarCode)':'obj_kodas',
    'Pavadinimas (name)':'name',
    'Savivaldybė, kurioje registruota(municipality)':'savivaldybe',
    'Ekonominės veiklos rūšies kodas(ecoActCode)':'veiklos_kodas',
    'Ekonominės veiklos rūšies pavadinimas(ecoActName)':'veiklos_aprasymas',
    'Mėnuo (month)':'date',
    'Vidutinis darbo užmokestis (avgWage)':'avgWage',
    'Apdraustųjų skaičius (numInsured)':'employee_count',
    'Vidutinis darbo užmokestis II (avgWage2)':'avgWage2',
    'Apdraustųjų skaičius II (numInsured2)':'employee_count2',
    'Valstybinio socialinio draudimo įmoka (tax)':'tax'
}

In [ ]:
# Setting DataFrames 
sod20 = pd.read_csv('data/monthly-2020.csv', sep=';')
sod21 = pd.read_csv('data/monthly-2021.csv', sep=';')

In [ ]:
# Remapping dataframes
sod20.rename(sod_map, axis=1, inplace=True)
sod21.rename(sod_map, axis=1, inplace=True)

In [ ]:
# Filtering data that does not contain obj_kodas since its the unique key that i will join data on. 
sod20 = sod20[sod20['obj_kodas'].notna()]
sod21 = sod21[sod21['obj_kodas'].notna()]

In [ ]:
# Ugly way to parse data, but it's what worked.
def to_dt(x):
    s = str(x)
    date = pd.to_datetime(s[:4] + '-' + s[-2:])
    return date

In [ ]:
# Applying dt parsing.
sod20['date'] = sod20['date'].apply(to_dt)
sod21['date'] = sod21['date'].apply(to_dt)

In [ ]:
# Employees
emp20 = sod20[sod20['date'].dt.month == 12].groupby(
    'obj_kodas')[['employee_count']].sum().reset_index().rename(columns={'employee_count': 'darbuotoju_skaicius_20'})
emp21 = sod21[sod21['date'].dt.month == 12].groupby(
    'obj_kodas')[['employee_count']].sum().reset_index().rename(columns={'employee_count': 'darbuotoju_skaicius_21'})
# Average wages
avg20 = sod20.groupby('obj_kodas')[['avgWage']].mean().reset_index().rename(
    columns={'avgWage': 'vid_atlyginimas_20'})
avg21 = sod21.groupby('obj_kodas')[['avgWage']].mean().reset_index().rename(
    columns={'avgWage': 'vid_atlyginimas_21'})
# information
info = sod21[['obj_kodas', 'savivaldybe', 'veiklos_kodas',
              'veiklos_aprasymas']].drop_duplicates()

features2 = [emp20, emp21, avg20, avg21, info]

In [ ]:
for feat in features2:
    sod = pd.merge(sod,feat,on='obj_kodas',how='left')

In [ ]:
sod = sod[cols]

## 2. Galutinis duomenų masyvas turi atrodyti kaip 3_uzduotis_data_sample.csv.

In [ ]:
sod.head(2)

## Užduotys

### 1. Išsiaiškinkite micro ir mažos įmonės apibrėžimą.

### 2. Remdamiesi apibrėžimu išfiltruokite micro ir mažas įmones. Įmonės turi turėti statusą MB arba UAB.

In [ ]:
sodf = sod[(
    sod['form_pav'].isin(['Uždaroji akcinė bendrovė', ' Mažoji bendrija'])) &
    (sod['pardavimo_pajamos_21'] < 10_000_000) & # Filtras skirtas Mažoms ir mikro įmonėms.
    (sod['pardavimo_pajamos_21'] > 10_000) & # Aktualios įmonės kurios turi bent 10k pajamų.
    (sod['pardavimo_pajamos_20'] > 10_000) & # Nufiltruojame mažo pelno įmones iš praitų metų kad negadintu 21 metų augimo.
    (sod['vid_atlyginimas_20'] > 642) & # Išfiltruojamos įmonės mokėjusios mažesnį nei minimalų atlygį.
    (sod['vid_atlyginimas_21'] < 10_000) & # Keletas įmonių turėjo net 30k atlyginimus 12mėn. Išfiltruojami bonusai.
    (sod['darbuotoju_skaicius_21'] < 50) # Filtras skirtas Mažoms ir mikro įmonėms.
].copy()

### 3. Indentifikuokite labiausiai augusias įmones.

In [ ]:
sodf['pardavimo_augimas_21'] = (sodf['pardavimo_pajamos_21'] - sodf['pardavimo_pajamos_20']) / sodf['pardavimo_pajamos_20'] * 100

In [ ]:
sodf['darbuotoju_augimas_21'] = (sodf['darbuotoju_skaicius_21'] - sodf['darbuotoju_skaicius_20']) / sodf['darbuotoju_skaicius_20'] * 100

In [ ]:
sodf['vid_atlyginimas_augimas_21'] = (sodf['vid_atlyginimas_21'] - sodf['vid_atlyginimas_20']) / sodf['vid_atlyginimas_20'] * 100

In [ ]:
top100sales =  sodf[['obj_pav','pardavimo_augimas_21']].replace(np.inf,0).sort_values(by = 'pardavimo_augimas_21', ascending=False).drop_duplicates().round()[:100]

In [ ]:
top100emps =  sodf[['obj_pav','darbuotoju_augimas_21']].replace(np.inf,0).sort_values(by = 'darbuotoju_augimas_21', ascending=False).drop_duplicates().round()[:100]

In [ ]:
top100wgs =  sodf[['obj_pav','vid_atlyginimas_augimas_21']].replace(np.inf,0).sort_values(by = 'vid_atlyginimas_augimas_21', ascending=False).drop_duplicates().round()[:100]

In [ ]:
pd.merge((pd.merge(top100sales,top100wgs,on='obj_pav',how='left')),top100emps,on='obj_pav',how='left')

### 4. Sukurkite kriterijus, ir atrinkite top 10, jūsų manymu geriausių įmonių.

In [ ]:
top100sales[:10]

### 5. Kokia suma įvertintumėt savo atrinktas įmones?

20 mln eur.

In [ ]:
#conda env export > environment.yml